commons/src/main/java/io/aiven/kafka/tieredstorage/commons/storage/s3/S3StorageConfig.java

```diff
@@ -0,0 +1,204 @@
+                new ConfigDef.NonEmptyString(),
+                ConfigDef.Importance.HIGH,
+                S3_BUCKET_NAME_DOC)
+            .define(
+                S3_ENDPOINT_URL_CONFIG,
+                ConfigDef.Type.STRING,
+                null,
+                ConfigDef.Importance.LOW,
+                S3_ENDPOINT_URL_DOC)
+            .define(
+                S3_REGION_CONFIG,
+                ConfigDef.Type.STRING,
+                S3_REGION_DEFAULT,
+                new RegionValidator(),
+                ConfigDef.Importance.MEDIUM,
+                S3_REGION_DOC)
+            .define(
+                S3_CREDENTIALS_PROVIDER_CLASS_CONFIG,
+                ConfigDef.Type.CLASS,
+                S3_CREDENTIALS_PROVIDER_CLASS_DEFAULT,
+                new CredentialsProviderValidator(),
+                ConfigDef.Importance.LOW,
+                S3_CREDENTIALS_PROVIDER_CLASS_DOC)
+            .define(
+                S3_ACCESS_KEY_ID_CONFIG,
+                ConfigDef.Type.PASSWORD,
+                null,
+                new NonEmptyPassword(),
+                ConfigDef.Importance.MEDIUM,
+                S3_ACCESS_KEY_ID_DOC)
+            .define(
+                S3_SECRET_ACCESS_KEY_CONFIG,
+                ConfigDef.Type.PASSWORD,
+                null,
+                new NonEmptyPassword(),
+                ConfigDef.Importance.MEDIUM,
+                S3_SECRET_ACCESS_KEY_DOC);
+    }
+
+    S3StorageConfig(final Map<String, ?> props) {
+        super(CONFIG, props);
+    }
+
+    AmazonS3 s3Client(final AWSCredentialsProvider credentialsProvider) {
+        AmazonS3ClientBuilder s3ClientBuilder = AmazonS3ClientBuilder.standard();
+        final String s3ServiceEndpoint = getString(S3_ENDPOINT_URL_CONFIG);
+        final String region = getString(S3_REGION_CONFIG);
+        if (s3ServiceEndpoint == null) {
+            s3ClientBuilder = s3ClientBuilder.withRegion(region);
+        } else {
+            final AwsClientBuilder.EndpointConfiguration endpointConfiguration =
+                new AwsClientBuilder.EndpointConfiguration(s3ServiceEndpoint, region);
+            s3ClientBuilder = s3ClientBuilder.withEndpointConfiguration(endpointConfiguration)
+                .withPathStyleAccessEnabled(true);
+        }
+        s3ClientBuilder.setCredentials(credentialsProvider);
+        return s3ClientBuilder.build();
+    }
+
+    public AWSCredentialsProvider credentialsProvider() {
+        try {
+            @SuppressWarnings("unchecked") final Class<? extends AWSCredentialsProvider> providerClass =
+                (Class<? extends AWSCredentialsProvider>) getClass(S3_CREDENTIALS_PROVIDER_CLASS_CONFIG);
+            final boolean areCredentialsProvided =
+                getPassword(S3_ACCESS_KEY_ID_CONFIG) != null && getPassword(S3_SECRET_ACCESS_KEY_CONFIG) != null;
+            if (areCredentialsProvided || providerClass.isAssignableFrom(AWSStaticCredentialsProvider.class)) {
```

Comment:
Since we instantiate concretely `AWSStaticCredentialsProvider` a few lines below, we cannot rely on assignability, it must be strict equals.
Maybe a better approach would be to disallow explicit credentials to be set at the same time with `s3.credentials.provider.class`: if the key is specified, the `s3.credentials.provider.class` must be not set and will be assumed as `AWSStaticCredentialsProvider`. Or we can require that if credentials are specified, the `s3.credentials.provider.class` must always be `AWSStaticCredentialsProvider` (i.e. same thing, but explicit). This would remove this confusion.
The global validation method I mentioned above could help with this.